core/src/sync/message/checkpoint_blame_state.rs

```diff
@@ -0,0 +1,166 @@
+        let trusted_block = ctx
+            .manager
+            .graph
+            .data_man
+            .block_header_by_hash(&self.trusted_blame_block);
+        if trusted_block.is_none() {
+            warn!(
+                "invalid trusted_blame_block={} in CheckpointBlameStateRequest from peer={}",
+                self.trusted_blame_block,
+                ctx.peer,
+            );
+            return Err(ErrorKind::Invalid.into());
+        }
+        let trusted_block = trusted_block.unwrap();
+
+        let mut state_blame_vec = Vec::new();
+        let mut block_hash = trusted_block.hash();
+        loop {
+            if let Some(exec_info) = ctx
+                .manager
+                .graph
+                .data_man
+                .consensus_graph_execution_info_from_db(&block_hash)
+            {
+                state_blame_vec.push(exec_info.original_deferred_state_root);
+                if state_blame_vec.len() == trusted_block.blame() as usize + 1 {
+                    break;
+                }
+                if let Some(block) =
+                    ctx.manager.graph.data_man.block_header_by_hash(&block_hash)
+                {
+                    block_hash = block.parent_hash().clone();
+                } else {
+                    warn!(
+                        "failed to find block={} in db, peer={}",
+                        block_hash, ctx.peer
+                    );
+                    return Err(ErrorKind::Invalid.into());
+                }
+            } else {
+                warn!("failed to find ConsensusGraphExecutionInfo for block={} in db, peer={}", block_hash, ctx.peer);
+                return Err(ErrorKind::Invalid.into());
+            }
+        }
+
+        let response = CheckpointBlameStateResponse {
+            request_id: self.request_id,
+            state_blame_vec,
+        };
+
+        ctx.send_response(&response)
+    }
+}
+
+impl Encodable for CheckpointBlameStateRequest {
+    fn rlp_append(&self, stream: &mut RlpStream) {
+        stream
+            .begin_list(2)
+            .append(&self.request_id)
+            .append(&self.trusted_blame_block);
+    }
+}
+
+impl Decodable for CheckpointBlameStateRequest {
+    fn decode(rlp: &Rlp) -> Result<Self, DecoderError> {
+        if rlp.item_count()? != 3 {
+            return Err(DecoderError::RlpIncorrectListLen);
+        }
+
+        Ok(CheckpointBlameStateRequest {
+            request_id: rlp.val_at(0)?,
+            trusted_blame_block: rlp.val_at(1)?,
+        })
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct CheckpointBlameStateResponse {
+    pub request_id: u64,
+    pub state_blame_vec: Vec<H256>,
+}
+
+build_msg_impl! { CheckpointBlameStateResponse, msgid::GET_CHECKPOINT_BLAME_STATE_RESPONSE, "CheckpointBlameStateResponse" }
+
+impl Handleable for CheckpointBlameStateResponse {
+    fn handle(self, ctx: &Context) -> Result<(), Error> {
+        ctx.match_request(self.request_id)?;
+
```

Comment:
Just double confirm, is it necessary to validate the response against the request in case of malicious node?